py3status/modules/speedtest.py

```diff
@@ -0,0 +1,273 @@
+# -*- coding: utf-8 -*-
+"""
+Do a bandwidth test with speedtest-cli.
+
+Configuration parameters:
+    button_share: button to open the result link (default None)
+    format: display format for this module, {download} and/or {upload} required
+        *(default 'Speedtest [\?color=darkgray ping [\?color=ping {ping} ms] '
+                    'up [\?color=upload {upload} {upload_unit}] '
+                    'down [\?color=download {download} {download_unit}]]')*
+    server_id: speedtest server to use, `speedtest-cli --list` to get id (default None)
+    si_units: use SI units (default False)
+    thresholds: specify color thresholds to use *(default {
+                'download': [(0, 'bad'), (1024, 'degraded'), (1024 * 1024, 'good')],
+                'ping': [(200, 'bad'), (150, 'orange'), (100, 'degraded'), (10, 'good')],
+                'upload': [(0, 'bad'), (1024, 'degraded'), (1024 * 1024, 'good')]})*
+    timeout: timeout when communicating with speedtest.net servers (default 10)
+    unit_bitrate: unit for download/upload rate (default 'MB/s')
+    unit_size: unit for bytes_received/bytes_sent (default 'MB')
+
+Format placeholders:
+    {bytes_sent}            bytes sent during test, human readable, eg '52.45'
+    {bytes_sent_unit}       unit used for bytes_sent, eg 'MB'
+    {bytes_sent_raw}        bytes sent during test, for format comparation, eg '52445184.0'
+    {bytes_received}        bytes received during test, human readable, eg '70.23'
+    {bytes_received_raw}    bytes received during test, for format comparation, eg '70229556.0'
+    {bytes_received_unit}   unit used for bytes_received, eg 'MB'
+    {client_country}        client country code, eg 'FR'
+    {client_ip}             client ip, eg '78.194.13.7'
+    {client_isp}            client isp, eg 'Free SAS'
+    {client_ispdlavg}       client isp download average, eg '0'
+    {client_isprating}      client isp rating, eg 3.7
+    {client_ispulavg}       client isp upload average, eg '0'
+    {client_lat}            client latitude, eg '48.8534'
+    {client_loggedin}       client logged in, eg '0'
+    {client_lon}            client longitude, eg '2.3487999999999998'
+    {client_rating}         client rating, eg '0'
+    {download}              download rate from speedtest server, human readable
+    {download_raw}          download rate from speedtest server, for format comparation
+    {download_unit}         unit used as , eg 'MB/s'
+    {ping}                  ping time in ms to speedtest server
+    {timestamp}             timestamp of the run, eg '2018-08-30T16:27:25.318212Z'
```

Comment:
One annoying thing about `Speedtest` is that if you refresh, you can't tell when it's finished. I think you should try make `{query_seconds}` or `{time_seconds}`. Maybe it can be part of default too next to `Speedtest`... so when we click it, we see the `[\?color=deepskyblue {seconds}]` going up... until it stopped there. That's when we know it's done. Lack of progress feedback sucks big time.  Just a thought. I guess `sleep_timeout = 0`.

Reply:
good idea, i thought on this too, will try

Reply:
Before I go further, I have done some try and I understand I need Threads to get working what we want. Before running command is a sort of long time polling command which block the rest of script.
If I run command in a Threads this should works.
We forget the idea or I go with Threads?

Reply:
Your code, your decision. We can forget this idea if you want to. 

Alternatively, see how it works when we empty the data on every `start()`. It should help.